### PR TITLE
Feature/codeception

### PR DESCRIPTION
**Pull request by guncha25**
Adds Codeception support:

1. Visual, Functional, Unit, Acceptance tests.
2. Per environment settings for tests.
3. Default PhantomJS browser load on test startup.
4. new command: ./build.sh test.

To test:
1. Build new drupal instalation
2. Make sure url's are correct in tests/suitname.suite.yml
3. From drupal directory run ./build.sh test

### CODE REVIEW

File: drupal/build.sh [hunk 1]
```diff
@@ -310,6 +310,24 @@ class Maker:
...
+                   codecept_args.extend([key, self.codecept_options[key]])
+               elif self.codecept_options[key]:
+                   codecept_args.extend([key])
+           if self.site_env != 'default' and self.site_env != 'local':
+               codecept_args.extend(["--env", self.site_env])
+           value = subprocess.call(codecept_args, cwd=web_dir) == 0
+           if not value:
+               raise BuildError("Tests failed.")
+       else:
+           self.warning("Needed packages are not required in this environment. Skipping tests.")
```
Comment (hugovk) [inline drupal/build.sh]: Can this be reworded? "Needed" and "required" are synonyms.

**Comment by floretan** [comment]
Even though build.sh is committed to this repository, it actually gets updated from a dedicated repository. We are in the process of relying less on it and using shell commands in site.yml rather than putting the code in build.sh, so I would definitely call `../vendor/bin/codecept` directly in site.yml rather than hide all the logic (and lose the options).

**Comment by guncha25** [comment]
@floretan Currently shell command doesn't fail build if command returns non-zero exit status. Question is - should it, but right now thats only reason why codecept command and options was added. Tests should fail build so that we can get report that something went wrong.

File: drupal/tests/functional.suite.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        - \Helper\Functional
+        - Asserts
+        - PhpBrowser:
+            url: http://local.wundertools.com
+        - DrupalBootstrap
+    config:
+        DrupalBootstrap:
+          drupal_root: '/vagrant/drupal/web'
+          site_path: 'sites/default'
+          ngnix_url: 'http://local.wundertools.com:8080'
```
Comment (onnimonni) [inline drupal/tests/functional.suite.yml]: Should this be `nginx_url`?

**Comment by misterjoonas** [comment]
I have tested this in our Finavia project, it seems to work after some composer issues.
But I can't use the username and password with getenv() which I have also merged, so I had to revert back to having credentials in settings.php:
  //'username' => getenv('DB_USER_DRUPAL'),
  //'password' => getenv('DB_PASS_DRUPAL'),
  'username' => 'drupal',
  'password' => 'password',

File: drupal/tests/functional.suite.yml (hunk 2 above)
Comment (misterjoonas) [inline drupal/tests/functional.suite.yml]: Fixed.

**Comment by floretan** [comment]
I would rather have the support steps and example tests upstream in https://github.com/guncha25/drupal-8-bootstrap, so that these can be updated after the start of the project, and this way we keep only a minimal codebase. A simple example could be provided, but I would really keep it as simple as possible (for example, go to the frontpage, I get a 200 response).

That can be moved later on though, I have no problem merging this branch as it is now.

**Comment by tharna** [comment]
Some documentation would be nice :)

**Pull request merged**